<launch>
    <!-- This work aims to further address the issues of SPH particle penetration and SPH systems navigating 
    through particularly narrow tunnels based on the foundation of Smoothed Particle Hydrodynamics (SPH), 
    with a focus on significantly improving the deformation capability of the SPH system.-->

    <arg name="particleCount" value="15"/>      <!-- 粒子的数量,确保你的内存足够大，一般要超过32G，16G最多支持20个左右的粒子 -->
    <arg name="particleInterval" value="0.8"/>  <!-- 粒子之间的初始间距 -->
    <arg name="particleVisScale" value="0.25"/> <!-- 粒子可视化尺寸 -->
    <arg name="sphInitTime" value="2.0"/>       <!-- 初始化时间 -->
    <arg name="init_bias_x" value="0.0"/>         <!-- 初始位置偏移量 -->
    <arg name="init_bias_y" value="0.0"/>         <!-- 初始位置偏移量 -->
    <arg name="updateInterval" value="0.02"/>   <!-- 时间步长，要与轨迹采样频率同步-->
    <arg name="h" value="15.0"/>                <!-- kernel function radius -->
    <arg name="mass" value="1.0"/>              <!-- default MASS -->
    <arg name="threshold_dist" value="0.03"/>   <!-- 粒子与虚拟粒子的碰撞检测距离 -->
    <arg name="k_den" value="1.0"/>           <!-- k_den -->
    <arg name="k_rep" value="1.0"/>           <!-- k_rep -->
    <arg name="k_fri" value="1.5"/>           <!-- k_fri -->
    <arg name="k_p"   value="6.0"/>           <!-- k_p -->
    <arg name="k_d"   value="3.5"/>           <!-- k_d -->
    <arg name="k_ff"  value="1.0"/>           <!-- k_ffs -->
    <arg name="k_force" value="100.0"/>       <!-- k_force -->
    <arg name="v_max" value="1.2"/>             <!-- v_max，必须大于规划速度-->
    <arg name="a_max" value="1.5"/>             <!-- a_max，必须大于规划加速度 -->
    <arg name="r_1"   value="0.06"/>            <!-- r_1 -->
    <arg name="r_2"   value="0.2"/>             <!-- r_2 -->
    <arg name="planv_max" value="0.8"/>             <!-- v_max -->
    <arg name="plana_max" value="1.0"/>             <!-- a_max -->
    <arg name="restDensity" value="1000000.0"/> <!-- \rho0 -->
    <arg name="g" value="-9.8"/>                <!-- 重力,no use -->
    <arg name="use_pctrl" value="true"/>        <!-- 使用位置\速度\加速度控制，默认使用位置控制 -->

    <!-- Run the sph_particles node -->
    <node name="sph_tase" pkg="water_swarm" type="sph_tase" output="screen">
        <param name="sph/mass"              value="$(arg mass)"/> 
        <param name="sph/restDensity"       value="$(arg restDensity)"/>
        <param name="sph/h"                 value="$(arg h)"/> 
        <param name="sph/g"                 value="$(arg g)"/>
        <param name="sph/use_pctrl"         value="$(arg use_pctrl)"/>
        <param name="sph/particleCount"     value="$(arg particleCount)"/> 
        <param name="sph/particleInterval"  value="$(arg particleInterval)"/>
        <param name="sph/updateInterval"    value="$(arg updateInterval)"/>
        <param name="sph/particleVisScale"  value="$(arg particleVisScale)"/>
        <param name="sph/threshold_dist"    value="$(arg threshold_dist)"/>
        <param name="sph/k_den"             value="$(arg k_den)"/> 
        <param name="sph/k_rep"             value="$(arg k_rep)"/> 
        <param name="sph/k_fri"             value="$(arg k_fri)"/> 
        <param name="sph/k_p"               value="$(arg k_p)"/> 
        <param name="sph/k_d"               value="$(arg k_d)" />
        <param name="sph/k_ff"              value="$(arg k_ff)" />
        <param name="sph/v_max"             value="$(arg v_max)"/>
        <param name="sph/a_max"             value="$(arg a_max)"/>
        <param name="sph/r_1"               value="$(arg r_1)"/>
        <param name="sph/r_2"               value="$(arg r_2)"/>
        <param name="sph/state_enabled"     value="false"/>
        <param name="sph/vis_role"          value="false"/>
        <param name="sph/init_bias_x"       value="$(arg init_bias_x)"/>
        <param name="sph/init_bias_y"       value="$(arg init_bias_y)"/>
    </node>

    <!-- Map-->
     <node pkg ="map_generator" name ="map_pub" type ="map_pub" output = "screen" args="$(find map_generator)/resource/tase_test.pcd"/>
    
    <!-- Navigation-->
    <node name="formation_planning" pkg="bspline_race" type="formation_planning" output="screen">
        <!-- Navigation param -->
        
    </node>

    <!-- RViz-->
    <node name="rviz" pkg="rviz" type="rviz" required="true" args="-d $(find so3_quadrotor_simulator)/config/experiment.rviz" />
    
</launch>